key_index.lua

```diff
@@ -0,0 +1,97 @@
+-- Storage to keep track of used keys. Allows to atomically create, delete
+-- and list keys. The keys are synchronized between nginx workers
+-- using ngx.shared.dict. The whole purpose of this module is to avoid
+-- using ngx.shared.dict:get_keys (see https://github.com/openresty/lua-nginx-module#ngxshareddictget_keys),
+-- which blocks all workers and therefore it shouldn't be used with large
+-- amounts of keys.
+
+local KeyIndex = {}
+KeyIndex.__index = KeyIndex
+
+local lock_lib = require("lock")
+
+function KeyIndex.new(shared_dict)
+  local self = setmetatable({}, KeyIndex)
+  self.dict = shared_dict
+  self.keys = {}
+  self.index = {}
+  self.lock = lock_lib.new("__lock_keys", self.dict)
+  return self
+end
+
+-- Loads new keys that might have been added by other workers since last sync.
+function KeyIndex:sync()
+  local N = self.dict:get("__key_count") or 0
+  -- Only sync if there are some new keys.
+  if N ~= #self.keys then
+    for i = #self.keys, N do
+      -- Read i-th key. If it is nil, it means it was deleted by some other thread.
+      local key = self.dict:get("__key_" .. i)
+      if key then
+        self.keys[i] = key
+        self.index[key] = i
+      end
+    end
+  end
+  return N
+end
+
+-- Returns list of all keys. Indices might contain "holes" in places where
+-- some keys were deleted.
+function KeyIndex:get()
+  self:sync()
+  return self.keys
+end
+
+-- Atomically adds one or more keys to the index.
+--
+-- Args:
+--   key_or_keys: Single string or a list of strings containing keys to add.
+function KeyIndex:add(key_or_keys)
+  local keys = key_or_keys
+  if type(key_or_keys) == "string" then
+    keys = { key_or_keys }
+  end
+
+  -- This must happen atomically, otherwise there could be a race condition
+  -- and other workers might create the same records at the same time
+  -- with different values.
+  if self.lock:wait() then
```

Comment:
I think instead of spinning to get a lock, workers can just spin trying to get a new unused index key. This should make locking unnecessary, I believe.

What are your thoughts on this?

```lua
for _, key in pairs(keys) do
  while (true) do
    local N = self:sync()
    if self.index[key] ~= nil then break end

    N = N+1
    ok, err = self.dict:safe_add(KEY_PREFIX .. N, key)
    if ok then
      self.dict:incr("__key_count", 1, 0)
      self.keys[N] = key
      self.index[key] = N
      break
    elseif err ~= "exists" then
      return "unexpected error adding a key: " .. err
    end
  end
end
```

Reply:
There is a small problem with this. Consider this (very weird, but possible) scenario:
 - worker A calls sync, which returns N=10
 - worker B calls sync, which returns N=10
 - worker A creates new metric (N+1=11)
 - worker A removes last metric (index[N=11] = nil)
 - worker B successfully calls safe_add with N=11
Now worker B thinks that the new metric has index 11, while worker A thinks it is deleted and could later create a new metric with the same name but different index.

It is actually somewhat similar to what @fffonion wrote below.

Reply:
Agreed, if the same worker creates and immediately removes the metric, other workers might be able to re-use the same index. Can we make this impossible by setting the index key to empty string instead of `nil` when a metric is deleted?

I'd really like to find a solution that does not involve our own lock implementation.

Reply:
I had the very same idea, just didn't have time to try it out yet. I would prefer lock-less solution as well.

Reply:
Implemented in cb9fbf22f13941269ead0e3e35453ab6d95c4975. After reading your suggestion few more times and more carefully, I believe it should actually work as is, since the deletions are now properly handled by `sync`, which is called before each attempt.

At first I didn't really like the fact that the `sync` is in the innermost loop, but it actually makes sense. It turns out to work kind of like [optimistic locking](https://en.wikipedia.org/wiki/Optimistic_concurrency_control), which is perfectly fine since we can assume that collisions will happen only very rarely.